Add explicit layout checks to Crossgen2

trylek opened the pull request:
Several tests under "Loader/classloader/explicitlayout" fail today
because Crossgen2 is missing GC consistency checks for types with
explicit layout - according to my understanding of the corresponding
CoreCLR algorithm we basically need to make sure that no GC fields
overlap non-GC fields. I have reimplemented the algorithm instead
of porting it as the CoreCLR version is heavily tied to the method
table builder and hard to separate out.

Thanks

Tomas

MichalStrehovsky commented:
This is rather difficult for me to follow with the `accessPath` and all. Do I understand it correctly that this goes over each field and for each field we check whether it overlaps any other field (by doing another foreach), basically doing n * n comparisons in the end?

I feel like this would be easier to follow (and more performant) if we do it like CoreCLR does it by allocating a byte array and setting bytes in it as we walk the fields in a single pass:

https://github.com/dotnet/coreclr/blob/af55c1b707d27a938c65742dc2522d88d7580108/src/vm/methodtablebuilder.cpp#L8324-L8328

I'm concerned that there's bugs in this around handling of pointer-typed fields (seems like we don't consider an overlap an error, but CoreCLR does):

```csharp
    [StructLayout(LayoutKind.Explicit)]
    unsafe struct A
    {
        [FieldOffset(0)] object o;
        [FieldOffset(0)] int* z;
    }
}
```

We also don't consider an overlap in padding an error, but CoreCLR does:

```csharp
    [StructLayout(LayoutKind.Explicit)]
    unsafe struct A
    {
        [FieldOffset(8)] object o;
    }

    [StructLayout(LayoutKind.Explicit)]
    unsafe struct B
    {
        [FieldOffset(0)] A a;
        [FieldOffset(0)] object o;
    }
```

Two more things:

* We should do this check before `ComputeExplicitFieldLayout` returns - this is a core type system invariant that we should check for and not just an R2R addon.
* The exception should match what CoreCLR does. `ExceptionStringID.ClassLoadExplicitLayout` is the right ID to use (it matches `IDS_CLASSLOAD_EXPLICIT_LAYOUT`), but the last parameter is an offset. We do this for compat - we want the exception messages to match what CoreCLR would throw. (The compat aspect is more pronounced in full AOT mode, but it's also here.)

Ideally, this would be accompanied by a unit test since it's a type system change and we need those to be well tested, but unfortunately we haven't ported unit tests to this repo :(

trylek commented:
@MichalStrehovsky - I believe I have addressed your PR feedback in the latest iteration; could you please take another look when you have a chance? At this point I haven't done anything about unit testing. The change fixes about 10~20 failures in existing CoreCLR tests under Loader\classloader\explicitlayout. Did you allude to porting some existing tests from the CoreRT repo?

MichalStrehovsky commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+                    }
+                }
+            }
+            else if (fieldType.IsPointer || fieldType.IsByRef || fieldType.IsByRefLike)
```
* IsByRefLike is a ref struct - those will respond to `IsValueType` above.
* We don't allow ByRef fields right now, so maybe we don't need to handle that.
* Pointer typed fields are okay to be misaligned. This works:

```csharp
[StructLayout(LayoutKind.Explicit)]
unsafe class Program
{
    [FieldOffset(2)] int* f;

    static unsafe void Main()
    {
    }
}
```

trylek replied:
Removed IsByRefLike from the conditional check. Added a new clause for IsPointer setting the range to NonORef.

MichalStrehovsky commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+            }
+            else if (fieldType.IsValueType)
+            {
+                MetadataType mdType = (MetadataType)fieldType;
```
If the field is a ByRef-like struct, it also needs to be aligned at pointer boundary. This should throw:

```csharp
[StructLayout(LayoutKind.Explicit)]
ref struct Program
{
    [FieldOffset(2)] Span<int> f;

    static unsafe void Main()
    {
    }
}
```

trylek replied:
Added alignment check for IsByRefLike within the IsValueType conditional block.

MichalStrehovsky commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+            }
+            else
+            {
+                throw new NotImplementedException();
```
Nit: Instead of an else branch that throws NotImplemented, I tend to use asserts. It doesn't make sense to ship the throw to customers and if we ever add new type system entities, the assert will hit.

trylek replied:
Changed NotImplementedException throw to assertion failure.

trylek commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+
+namespace Internal.TypeSystem
+{
+    public class ExplicitLayoutValidator
```
Of course, thanks for pointing that out!

trylek commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+
+using Internal.TypeSystem;
+using System;
+using System.Net;
```
No use, I just accidentally copied that over along with the copyright header. Deleted.

trylek commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,146 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.TypeSystem;
```
Deleted.

trylek commented:
OK, that should be it (modulo the tests I haven't done anything about at this point), @MichalStrehovsky - could you please take another look?

MichalStrehovsky commented on src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs:
```diff
@@ -0,0 +1,155 @@
+                }
+                SetFieldLayout(offset, _pointerSize, FieldLayoutTag.ORef);
+            }
+            else if (fieldType.IsPointer)
```
```suggestion
            else if (fieldType.IsPointer || fieldType.IsFunctionPointer)
```

trylek replied:
Fixed in 3rd commit.
